File: apps/server/internal/dtos/s3.go
package dtos

import (
	"regexp"
	"time"

	"cloudrive/server/internal/lib/validator"

	"github.com/google/uuid"
)

// bucketNamePattern mirrors the S3 bucket naming rules enforced by the
// database CHECK constraint (3-63 chars, lowercase, dots/dashes).
var bucketNamePattern = regexp.MustCompile(`^[a-z0-9][a-z0-9.-]{1,61}[a-z0-9]$`)

type CreateS3CredentialRequest struct {
	Label string `json:"label"`
}

func (dto CreateS3CredentialRequest) Validate(v *validator.MapValidator) {
	v.Field("label").String()
}

type S3CredentialResponse struct {
	ID          uuid.UUID  `json:"id"`
	WorkspaceID uuid.UUID  `json:"workspace_id"`
	AccessKeyID string     `json:"access_key_id"`
	SecretKey   string     `json:"secret_key,omitempty"` // only on creation
	Label       *string    `json:"label,omitempty"`
	Status      string     `json:"status"`
	LastUsedAt  *time.Time `json:"last_used_at,omitempty"`
	CreatedAt   time.Time  `json:"created_at"`
}

type CreateS3BucketRequest struct {
	Name             string `json:"name"`
	StorageAccountID string `json:"storage_account_id"`
	RootPrefix       string `json:"root_prefix"`
}

func (dto CreateS3BucketRequest) Validate(v *validator.MapValidator) {
	v.Field("name").Required().Match(bucketNamePattern)
	v.Field("storage_account_id").Required().UUID()
	v.Field("root_prefix").String()
}

File: apps/app/lib/api/models/s3.ts
import type { ISO8601DateString } from '@/types/time'

// Mirrors dtos.S3CredentialResponse — secret_key is present only on creation.
export interface S3Credential {
  id: string
  workspace_id: string
  access_key_id: string
  secret_key?: string
  label?: string
  status: string
  last_used_at?: ISO8601DateString | null
  created_at: ISO8601DateString
}

export interface S3Bucket {
  id: string
  organization_id: string
  workspace_id: string
  name: string
  storage_account_id: string
  root_prefix: string
  created_by: string
  created_at: ISO8601DateString
  updated_at: ISO8601DateString
}
